src/backend/executor/nodeModifyTable.c

```diff
@@ -2506,3 +2448,69 @@ ExecSquelchModifyTable(ModifyTableState *node)
 			break;
 	}
 }
+
+/*
+ * Build a attribute mapping between child partition and the root partition in
+ * case if child partition has physically-different attribute numbers from
+ * root's due to dropped columns.
+ */
+void
+makePartitionCheckMap(EState *estate, ResultRelInfo *resultRelInfo)
+{
+	Relation	resultRelationDesc = resultRelInfo->ri_RelationDesc;
+	Oid			parentRelid;
+
+	Assert(PointerIsValid(estate->es_result_partitions));
+
+	/*
+	 * The partition selection operation is done with the parent partition's
+	 * attribute numbers, so if child partition has physically-different
+	 * attribute numbers due to dropped columns, we should map the child
+	 * attribute numbers to the parent's attribute numbers to perform the
+	 * partition selection. EState may not have the parent relation
+	 * information at the moment, so we have to do a hard job here by opening
+	 * it and compare the tuple descriptors. The purpose of this code is to
+	 * prevent the tuple from being incorrectly interpreted during partition
+	 * selection, that can be performed in ExecInsert, ExecDelete and
+	 * checkPartitionUpdate functions when we work with the leaf partition as
+	 * result relation.
+	 */
+	parentRelid = estate->es_result_partitions->part->parrelid;
+
+	if (resultRelInfo->ri_PartCheckTupDescMatch == 0 &&
+		parentRelid != RelationGetRelid(resultRelInfo->ri_RelationDesc))
+	{
+		Relation	parentRel;
+		TupleDesc	resultTupdesc,
+					parentTupdesc;
+
+		/*
+		 * We are on a child part, let's see the tuple descriptor looks like
+		 * the parent's one.  Probably this won't cause deadlock because DML
+		 * should have opened the parent table with appropriate lock.
+		 */
+		parentRel = relation_open(parentRelid, AccessShareLock);
+		resultTupdesc = RelationGetDescr(resultRelationDesc);
+		parentTupdesc = RelationGetDescr(parentRel);
+		if (!equalTupleDescs(resultTupdesc, parentTupdesc, false))
+		{
+			AttrMap    *map;
+			MemoryContext oldcontext;
+
+			/* Tuple looks different. Construct attribute mapping. */
+			oldcontext = MemoryContextSwitchTo(estate->es_query_cxt);
+			map_part_attrs(resultRelationDesc, parentRel, &map, true);
+			MemoryContextSwitchTo(oldcontext);
+
+			/* And save it for later use. */
+			resultRelInfo->ri_PartCheckMap = map;
+
+			resultRelInfo->ri_PartCheckTupDescMatch = -1;
+		}
+		else
+			resultRelInfo->ri_PartCheckTupDescMatch = 1;
```

Comment:
Why not all comments and code were moved here from checkPartitionUpdate? For example
```c
resultRelInfo->ri_PartCheckTupDescMatch = 0;
```
and/or
```c
	/*
	 * Check this at the first pass only to avoid repeated catalog access.
	 */
```
and so on

Reply:
Have resotred omitted code and comments